File: 0152-maximum-product-subarray/0152-maximum-product-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */


var maxProduct = function(nums) {

    // initialize the res variable
    let res = Math.max(...nums);
    
    let currMin = 1;
    let currMax = 1;
    
    // loop through the nums array:
    for (let num of nums) {
        
        // edge case:0
        if (num == 0) {
            
            currMin = currMax = 1;
            continue;
        }
        
        let temp = currMax;
        
        currMax = Math.max(currMax*num, currMin*num, num);
        currMin = Math.min(temp*num, currMin*num, num);
        res = Math.max(res, currMax);
    }
    
    return res;
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
// find the max. product subArray; subArray has to be continuous

// approach: loop through nums once; for each loop, find max(n, currMin, currMax), return res in the end

class Solution {
    public int maxProduct(int[] nums) {
     
        // init res to the max. number of nums:
        int res = Integer.MIN_VALUE;
        for (int number:nums) {
            res = Math.max(res, number);
        }
        
        // init 2 global variabes: currMin currMax
        int currMin = 1;
        int currMax = 1;
        
        // loop through nums
        for (int i=0; i<nums.length; i++) {
            
            // edge case:0 - reset currMin & currMax
            if (nums[i] == 0) {
                currMin = 1;
                currMax = 1;
                continue;
            }
            
            // compute:
            int temp = currMin;
            currMin = Math.min(currMin * nums[i], Math.min(currMax * nums[i], nums[i]));
            currMax = Math.max(temp * nums[i], Math.max(currMax * nums[i], nums[i]));
            
            // update res:
            // res = Math.max(res, currMin, currMax);
            res = Collections.max(Arrays.asList(res, currMin, currMax));
        }
        
        return res;
    }
}